add no battery mode for mainboards without battery

saizo80 opened the pull request:
if using a mainboard without a battery, like in a mounted case or as a server, ectool will error when fetching temps on the battery sensor which can be seen with this command.

```bash
ectool tempsinfo all
0: 1 F75303_Local
1: 0 F75303_CPU
2: 1 F75303_DDR
3: 3 Battery
4: 0 PECI
```

system journal errors:

```bash
Aug 18 10:04:00 hostname python3[346601]: Sensor 3 error
Aug 18 10:04:01 hostname python3[346608]: Sensor 3 error
Aug 18 10:04:02 hostname python3[346611]: Sensor 3 error
Aug 18 10:04:03 hostname python3[346614]: Sensor 3 error
Aug 18 10:04:04 hostname python3[346620]: Sensor 3 error
Aug 18 10:04:05 hostname python3[346631]: Sensor 3 error
```

this will cause an error message in journalctl once every second when fw-fanctrl calls ectool. this pr adds a config option `noBatteryMode` which will only call ectool on every sensor except for one marked as battery.

leopoldhub commented:
Hi, thanks for pointing this out and for your contribution.

I'll have a good look at everything in a bit, but here is my initial feedback.

At the moment, when we deal with instance configurations, we do send them via the run command (e.g. hardware-controller, socket-controller, silent...).
I am open to imagining a new standard way of configuring this via the configuration file, but to ensure uniformity, I think we should do it this way for now. (feel free to discuss this in a new issue 😉)

I will be available in about 2 hours to give you more feedback.
If you have any questions, feel free to ask them and I will do my best to answer.

saizo80 commented:
That's no problem, I don't think it'll be too much work to rework it to be a run flag like that. I'll work on it later today

saizo80 commented:
i've made changes that pull from the command line args for the no battery mode, and also the changes to the installer to add the flag to the ExecStart in the service file

leopoldhub commented on fanctrl.py:
```diff
@@ -81,6 +81,11 @@ def initParser(self):
             )
+            runCommand.add_argument(
+                "--no-battery",
+                help="disable ectool checking the battery sensor",
```
Maybe should we make the "no-battery" option more general and not ectool specific?

leopoldhub commented on fanctrl.py:
```diff
@@ -229,6 +238,8 @@ def parseArgs(self, args=None):
                 values.strategy = legacy_values.strategy
             if not hasattr(values, "command"):
                 raise Exception("not a valid legacy command")
+            if not hasattr(values, "no_battery") and legacy_values.no_battery is not None:
```
same here, it is not needed in the legacy parser

leopoldhub commented on fanctrl.py:
```diff
@@ -463,20 +486,36 @@ class FanController:
     tempHistory = collections.deque([0] * 100, maxlen=100)
     active = True
     timecount = 0
+    noBatteryMode = False
```
Perhaps the "noBatteryMode" should be in the hardware controller instead, as it's only purpose is to interface with the hardware to parse temperature and fan speed?

leopoldhub reviewed (commented):
Hi, here are more in-depth feedbacks.

If you disagree or have arguments to defend, feel free to say so 😉.

I may be nitpicking, but what do you think about naming the argument "no-battery-sensor" instead?
I am afraid that "no-battery" might be understood as "disable all battery related features", which includes the profile switch between "on ac" and "on battery".

Otherwise great work!

saizo80 commented:
no i think these are all valid things to point out. i've updated things accordingly, and moved most of the work to the hardware controller. the only thing i can't really speak for is using a different hardware controller other than ectool, because there is not other code here to use such, and i only know the output format of ectool. so these class specific modifications are probably a good way to go in case another controller is added in the future

leopoldhub commented:
> the only thing i can't really speak for is using a different hardware controller other than ectool, because there is not other code here to use such, and i only know the output format of ectool. so these class specific modifications are probably a good way to go in case another controller is added in the future

Yes, this is just a "what if", but it helps to structure the code and assign responsibilities to each class.

I just have one last bit of feedback to give you and we should be good to merge 👍

leopoldhub commented on fanctrl.py:
```diff
@@ -429,9 +434,31 @@ def isOnAC(self):
+        if self.noBatteryMode:
+            self.getNonBatterySensors()
+    
+    def getNonBatterySensors(self):
```
When a function is named "getxxxx", I expect a value to be returned and that no global (or class in this case) variable assignment is made.
Perhaps we should instead assign the "nonBatterySensors" in the init or in an other accordingly named method and return the filtered values?

saizo80 replied:
i don't much like having so much much code in the init function, so would the function name `populateNonBatterySensors` be alright for this?

leopoldhub commented on fanctrl.py:
```diff
@@ -429,9 +434,31 @@ def isOnAC(self):
 
 
 class EctoolHardwareController(HardwareController, ABC):
+    noBatteryMode = False
```
"noBatterySensorMode"?

leopoldhub commented on install.sh:
```diff
@@ -24,6 +24,7 @@ SHOULD_INSTALL_ECTOOL=true
 SHOULD_PRE_UNINSTALL=true
 SHOULD_POST_INSTALL=true
 SHOULD_REMOVE=false
+NO_BATTERY=false
```
"NO_BATTERY_SENSOR"?

leopoldhub commented:
oops, one last thing, there is no documentation for "--no-battery-sensors" the readme for the run command

leopoldhub commented:
After some more testing, it seems that the way you add the "--no-battery-sensors" to the service will not do :/

If I run the install script once with the "--no-battery-sensors" argument, then uninstall and reinstall again without it, it still stays in the service anyway.

The integration of the argument into the service should be done without persistence in the installation files.
You could use the actual way we do this by using a placeholder and replacing it with the desired value (Line 166 `cat "$SERVICES_DIR/$SERVICE$SERVICE_EXTENSION" | sed -e "s/%PREFIX_DIRECTORY%/${PREFIX_DIR//\//\\/}/" ..... | tee "$DEST_DIR$PREFIX_DIR/lib/systemd/system/$SERVICE$SERVICE_EXTENSION"`).

saizo80 commented:
fixed the installer and renamed a lot of variables/functions for clarity in these commits

leopoldhub reviewed (approved):
Perfect, everything seems good to me 👍

Do you want to change anything else before the merge?

saizo80 commented:
nope, it all looks good to me!

leopoldhub commented:
🎉 Congratulations 🎉

Thank you again for your time and contribution!
It was a pleasure to work with you.

Have a great day :)
